crates/cdk/src/publish.rs

```diff
@@ -59,70 +64,76 @@ pub struct PublishCmd {
-        let hubdir = package_info.package_relative_path(DEF_HUB_INIT_DIR);
-
-        info!(
-            "publishing package {} from {}",
-            package_info.package_name(),
-            package_info.package_path().to_string_lossy()
-        );
-
-        self.cleanup(&package_info)?;
-
-        if !self.no_build {
-            build_connector(&package_info, BuildOpts::with_release(opt.release.as_str()))?;
-        }
-
-        init_package_template(&package_info, &self.readme)?;
-        check_package_meta_visiblity(&package_info)?;
 
         match (self.pack, self.push) {
             (false, false) | (true, true) => {
-                let pkgmetapath = self
-                    .package_meta
-                    .as_ref()
-                    .map(PathBuf::from)
-                    .unwrap_or_else(|| hubdir.join(hubutil::HUB_PACKAGE_META));
+                let opt = self.package.as_opt();
+                let hubdir = self.run_in_cargo_project(&opt)?;
+                let pkgmetapath = self.package_meta_path(&hubdir);
                 let pkgdata = package_assemble(pkgmetapath, &opt.target, &access)?;
                 package_push(self, &pkgdata, &access)?;
+                self.cleanup(&hubdir)?;
             }
 
             // --pack only
             (true, false) => {
-                let pkgmetapath = self
-                    .package_meta
-                    .as_ref()
-                    .map(PathBuf::from)
-                    .unwrap_or_else(|| hubdir.join(hubutil::HUB_PACKAGE_META));
+                let opt = self.package.as_opt();
+                let hubdir = self.run_in_cargo_project(&opt)?;
+                let pkgmetapath = self.package_meta_path(&hubdir);
                 package_assemble(pkgmetapath, &opt.target, &access)?;
+                self.cleanup(&hubdir)?;
             }
 
             // --push only, needs ipkg file
             (false, true) => {
-                let pkgfile = &self
-                    .package_meta
-                    .clone()
-                    .ok_or_else(|| anyhow::anyhow!("package file required for push"))?;
+                let pkgfile = self.ipkg_file.as_ref().ok_or_else(|| {
+                    anyhow::anyhow!("need to specify --ipkg-file when using --push")
+                })?;
                 package_push(self, pkgfile, &access)?;
             }
         }
 
-        if !self.pack {
-            self.cleanup(&package_info)?;
-        }
-
         Ok(())
     }
 
-    pub fn cleanup(&self, package_info: &PackageInfo) -> Result<()> {
+    fn package_meta_path(&self, hubdir: &Path) -> PathBuf {
+        self.package_meta
+            .as_ref()
+            .map(PathBuf::from)
+            .unwrap_or_else(|| hubdir.join(hubutil::HUB_PACKAGE_META))
+    }
+
+    // This gets run only if the command should be run in the cargo project folder
+    // of the connector
+    //
+    // returns hubdir
+    fn run_in_cargo_project(&self, opt: &PackageOption) -> Result<PathBuf> {
```

Comment:
The function is called `run_in_cargo_project` but it does: cleanup, building connector, init package, check visibility.

Reply:
I was thinking more of having high level methods `push`, `pack` and `build` and hiding specific things inside.

Reply:
it is more of a `if` so this gets run if we are running in a cargo project

Reply:
high level methods don't really work without this function because we want to run these in two cases and not run them on one (only --push) case. High level methods would be the statements in the match expression

Reply:
Ok, let's return to your first version but instead of `unreachable` just print and return an error.

Reply:
We can come up with a new structure later, for now, let's just fix a bug with minimal refactoring.

Reply:
tbh, I don't want to revert because imo this is cleaner and also reverting will take more time

Reply:
Again, a function that does a lot of things, and returns a path to hub directory and with name `run_in_cargo_project` is very far from clean and readable code.

Reply:
It is debatable if it is cleaner than putting the if at the start of the function and having a needless arm in match that references the if at the end of the function. It is code to delete/create folders and make http requests in the end don't need to work days to make it perfect